File: Projetos/ACBrLib/Demos/C#/Reinf/Import/Estático/ACBrReinf.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using ACBrLib.Core;
using ACBrLib.Core.DFe;

namespace ACBrLib.Reinf
{
    /// <inheritdoc />
    public sealed partial class ACBrReinf : ACBrLibHandle
    {
        #region Constructors

        public ACBrReinf(string eArqConfig = "", string eChaveCrypt = "") : base(IsWindows ? "ACBrReinf64.dll" : "libacbrReinf64.so",
                                                                                      IsWindows ? "ACBrReinf32.dll" : "libacbrReinf32.so")
        {
            var inicializar = GetMethod<Reinf_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);

            Config = new ACBrReinfConfig(this);
        }

        #endregion Constructors

        #region Properties

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Reinf_Nome>();
                var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Reinf_Versao>();
                var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public ACBrReinfConfig Config { get; }

        #endregion Properties

        #region Metodos

        #region Ini

        public override void ConfigGravar(string eArqConfig = "")
        {
            var gravarIni = GetMethod<Reinf_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override void ImportarConfig(string eArqConfig)
        {
            var lerIni = GetMethod<Reinf_ConfigImportar>();
            var ret = ExecuteMethod(() => lerIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override string ExportarConfig()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Reinf_ConfigExportar>();
            var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public override void ConfigLer(string eArqConfig = "")
        {
            var lerIni = GetMethod<Reinf_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<Reinf_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public override void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<Reinf_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        #endregion Ini

        #region Diversos
        public void CriarEventoReinf (string eArqIni)
        {
            var method = GetMethod<Reinf_CriarEventoReinf>();
            var ret = ExecuteMethod(() => method(ToUTF8(eArqIni)));

            CheckResult(ret);
        }

        public string EnviarReinf()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Reinf_EnviarReinf>();
            var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string ConsultarReinf(string eProtocolo)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Reinf_ConsultarReinf>();
            var ret = ExecuteMethod(() => method(ToUTF8(eProtocolo), buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }
        public string ConsultarReciboReinf(string ePerApur, int aTipoEvento, string eNrInscEstab,
            string eCnpjPrestador, string eNrInscTomador, string eDtApur, string eCpfCnpjBenef,
            string eCnpjFonte)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Reinf_ConsultarReciboReinf>();
            var ret = ExecuteMethod(() => method(ToUTF8(ePerApur), aTipoEvento, ToUTF8(eNrInscEstab),
                                                 ToUTF8(eCnpjPrestador), ToUTF8(eNrInscTomador), 
                                                 ToUTF8(eDtApur), ToUTF8(eCpfCnpjBenef),
                                                 ToUTF8(eCnpjFonte),
                                                 buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string CriarEnviarReinf(string eArqIni)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Reinf_CriarEnviarReinf>();
            var ret = ExecuteMethod(() => method(ToUTF8(eArqIni), buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public void LimparReinf()
        {
            var method = GetMethod<Reinf_LimparReinf>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void CarregarXMLEventoReinf(string eArquivoOuXML)
        {
            var method = GetMethod<Reinf_CarregarXMLEventoReinf>();
            var ret = ExecuteMethod(() => method(ToUTF8(eArquivoOuXML)));

            CheckResult(ret);
        }

        public void SetIDContribuinte(string aIdContribuinte)
        {
            var method = GetMethod<Reinf_SetIDContribuinte>();
            var ret = ExecuteMethod(() => method(ToUTF8(aIdContribuinte)));

            CheckResult(ret);
        }

        public void SetIDTransmissor(string aIdTransmissor)
        {
            var method = GetMethod<Reinf_SetIDTransmissor>();
            var ret = ExecuteMethod(() => method(ToUTF8(aIdTransmissor)));

            CheckResult(ret);
        }

        public void TipoContribuinte(int aTipoContribuinte)
        {
            var method = GetMethod<Reinf_SetTipoContribuinte>();
            var ret = ExecuteMethod(() => method(aTipoContribuinte));

            CheckResult(ret);
        }

        public void SetVersao(string sVersao)
        {
            var method = GetMethod<Reinf_SetVersaoDF>();
            var ret = ExecuteMethod(() => method(ToUTF8(sVersao)));

            CheckResult(ret);
        }

        public InfoCertificado[] ObterCertificados()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Reinf_ObterCertificados>();
            var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            var certificados = ProcessResult(buffer, bufferLen).Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
            return certificados.Length == 0 ? new InfoCertificado[0] : certificados.Select(x => new InfoCertificado(x)).ToArray();
        }

        #endregion Diversos

        #region Private Methods

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<Reinf_Finalizar>();
            var codRet = ExecuteMethod(() => finalizar());
            CheckResult(codRet);
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<Reinf_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));
            return FromUTF8(buffer);
        }

        #endregion Private Methods

        #endregion Metodos
    }
}

File: Projetos/ACBrLib/Demos/Java/Reinf/Imports/ACBrLibReinfMT/src/com/acbr/reinf/ACBrReinf.java
package com.acbr.reinf;

import com.acbr.ACBrLibBase;
import com.acbr.ACBrSessao;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;

import java.io.File;
import java.nio.ByteBuffer;
import java.nio.file.Paths;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;

//lib MT

public final class ACBrReinf extends ACBrLibBase {

    private interface ACBrReinfLib extends Library {

        static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public final static ACBrReinfLib INSTANCE = LibraryLoader.getInstance();

        class LibraryLoader {

            private static String library = "";
            private static ACBrReinfLib instance = null;

            private static String getLibraryName() {
                if (library.isEmpty()) {
                    if (Platform.isWindows()) {
                        library = Platform.is64Bit() ? "ACBrReinf64" : "ACBrReinf32";
                    } else {
                        library = Platform.is64Bit() ? "ACBrReinf64" : "ACBrReinf32";
                    }
                }
                return library;
            }

            public static ACBrReinfLib getInstance() {
                if (instance == null) {
                    instance = (ACBrReinfLib) Native.synchronizedLibrary(
                            (Library) Native.load(JNA_LIBRARY_NAME, ACBrReinfLib.class));
                }

                return instance;
            }
        }

        int Reinf_Inicializar(PointerByReference libHandler, String eArqConfig, String eChaveCrypt);

        int Reinf_Finalizar(Pointer libHandler);

        int Reinf_Nome(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Reinf_Versao(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Reinf_UltimoRetorno(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Reinf_ConfigImportar(Pointer libHandler, String eArqConfig);

        int Reinf_ConfigExportar(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Reinf_ConfigLer(Pointer libHandler, String eArqConfig);

        int Reinf_ConfigGravar(Pointer libHandler, String eArqConfig);

        int Reinf_ConfigLerValor(Pointer libHandler, String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);

        int Reinf_ConfigGravarValor(Pointer libHandler, String eSessao, String eChave, String valor);

        int Reinf_CriarEventoReinf(Pointer libHandler, String eArqIni);
        
        int Reinf_EnviarReinf(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
        
        int Reinf_ConsultarReinf(Pointer libHandler, String eProtocolo, ByteBuffer buffer, IntByReference bufferSize);
    
        int Reinf_ConsultarReciboReinf(Pointer libHandler, String ePerApur, int aTipoEvento, String eNrInscEstab, 
                String eCnpjPrestador, String eNrInscTomador, String eDtApur, 
                String eCpfCnpjBenef, String eCnpjFonte, 
                ByteBuffer buffer, IntByReference bufferSize);
       
        int Reinf_CriarEnviarReinf(Pointer libHandler, String eArqIni, ByteBuffer buffer, IntByReference bufferSize);
        
        int Reinf_LimparReinf(Pointer libHandler);
        
        int Reinf_CarregarXMLEventoReinf(Pointer libHandler, String eArquivoOuXML);
        
        int Reinf_SetIDContribuinte(Pointer libHandler, String aIdContribuinte);
        
        int Reinf_SetIDTransmissor(Pointer libHandler, String aIdTransmissor);
        
        int Reinf_SetTipoContribuinte(Pointer libHandler, Integer aTipoContribuinte);
        
        int Reinf_SetVersaoDF(Pointer libHandler, String sVersao);
        
        int Reinf_ObterCertificados(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
    }

    public ACBrReinf() throws Exception {
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if (!iniFile.exists()) {
            iniFile.createNewFile();
        }

        PointerByReference handle = new PointerByReference();
        int ret = ACBrReinfLib.INSTANCE.Reinf_Inicializar(handle, toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
        setHandle(handle.getValue());
    }

    public ACBrReinf(String eArqConfig, String eChaveCrypt) throws Exception {
        PointerByReference handle = new PointerByReference();
        int ret = ACBrReinfLib.INSTANCE.Reinf_Inicializar(handle, toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
        setHandle(handle.getValue());
    }

    @Override
    protected void dispose() throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_Finalizar(getHandle());
        checkResult(ret);
    }

    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrReinfLib.INSTANCE.Reinf_Nome(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrReinfLib.INSTANCE.Reinf_Versao(getHandle(),buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_ConfigLer(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_ConfigGravar(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    @Override
    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrReinfLib.INSTANCE.Reinf_ConfigLerValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    @Override
    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_ConfigGravarValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }

    public void ConfigImportar(String eArqConfig) throws Exception {

        int ret = ACBrReinfLib.INSTANCE.Reinf_ConfigImportar(getHandle(), eArqConfig);
        checkResult(ret);
    }

    public String ConfigExportar() throws Exception {

        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrReinfLib.INSTANCE.Reinf_ConfigExportar(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }
    
    public void criarEventoReinf(String eArqIni) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_CriarEventoReinf(getHandle(), eArqIni);
        checkResult(ret);
    }
    
    public String enviarReinf() throws Exception {
        // Buffer deve ser maior que 256 para suportar o retorno completo
        ByteBuffer buffer = ByteBuffer.allocate(16384);
        IntByReference bufferLen = new IntByReference(16384);
        
        int ret = ACBrReinfLib.INSTANCE.Reinf_EnviarReinf(getHandle(), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    public String consultarReinf(String eProtocolo) throws Exception {
        // Buffer deve ser maior que 256 para suportar o retorno completo
        ByteBuffer buffer = ByteBuffer.allocate(16384);
        IntByReference bufferLen = new IntByReference(16384);
        
        int ret = ACBrReinfLib.INSTANCE.Reinf_ConsultarReinf(getHandle(), eProtocolo, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    public String consultarReciboReinf(String ePerApur, int aTipoEvento, String eNrInscEstab, 
                String eCnpjPrestador, String eNrInscTomador, String eDtApur, 
                String eCpfCnpjBenef, String eCnpjFonte) throws Exception {
        
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
        
        int ret = ACBrReinfLib.INSTANCE.Reinf_ConsultarReciboReinf(getHandle(), ePerApur, aTipoEvento, eNrInscEstab, 
                eCnpjPrestador, eNrInscTomador, eDtApur, 
                eCpfCnpjBenef, eCnpjFonte, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public String criarEnviarReinf(String eArqIni) throws Exception {
        // Buffer deve ser maior que 256 para suportar o retorno completo
        ByteBuffer buffer = ByteBuffer.allocate(16384);
        IntByReference bufferLen = new IntByReference(16384);

        int ret = ACBrReinfLib.INSTANCE.Reinf_CriarEnviarReinf(getHandle(), eArqIni, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    public void limparReinf() throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_LimparReinf(getHandle());
        checkResult(ret);
    }
    
    public void carregarXMLEventoReinf(String eArquivoOuXML) throws Exception{
        int ret = ACBrReinfLib.INSTANCE.Reinf_CarregarXMLEventoReinf(getHandle(), eArquivoOuXML);
        checkResult(ret);
    }
    
    public void setIDContribuinte(String aIdContribuinte) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_SetIDContribuinte(getHandle(), aIdContribuinte);
        checkResult(ret);
    }
    
    public void setIDTransmissor(String aIdTransmissor) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_SetIDTransmissor(getHandle(), aIdTransmissor);
        checkResult(ret);
    }
    
    public void setTipoContribuinte(Integer aTipoContribuinte) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_SetTipoContribuinte(getHandle(), aTipoContribuinte);
        checkResult(ret);
    }
    
    public void setVersaoDF(String sVersao) throws Exception {
        int ret = ACBrReinfLib.INSTANCE.Reinf_SetVersaoDF(getHandle(), sVersao);
        checkResult(ret);
    }
    
    public String obterCertificados() throws Exception {
      ByteBuffer buffer = ByteBuffer.allocate( STR_BUFFER_LEN );
      IntByReference bufferLen = new IntByReference( STR_BUFFER_LEN );
      
      int ret = ACBrReinfLib.INSTANCE.Reinf_ObterCertificados(getHandle(), buffer, bufferLen );
      checkResult(ret);
      return processResult( buffer, bufferLen );
    }
    
    @Override
    protected void UltimoRetorno(ByteBuffer buffer, IntByReference bufferLen) {
        ACBrReinfLib.INSTANCE.Reinf_UltimoRetorno(getHandle(), buffer, bufferLen);
    }
}
